sycl/doc/SYCL2020-SpecializationConstants.md

```diff
@@ -0,0 +1,766 @@
+# Specialization constants
+
+Specialization constants are implemented in accordance with how they are defined
+by SYCL 2020 specification: [SYCL registry][sycl-registry],
+[direct link to the specification][sycl-2020-spec].
+
+[sycl-registry]: https://www.khronos.org/registry/SYCL/
+[sycl-2020-spec]: https://www.khronos.org/registry/SYCL/specs/sycl-2020/pdf/sycl-2020.pdf
+
+TODO: feature overview? code example?
+
+## Design
+
+[SYCL 2020][sycl-2020-spec] defines specialization constant as:
+
+> A constant variable where the value is not known until compilation of the
+> SYCL kernel function.
+>
```

Comment:
Can we get clarification about exactly what constitutes compilation time of the SYCL kernel function?
In ahead of time compilation does this mean that all specialization constants must be known at that time?
In JIT compilation, does this mean the the first time the kernel is jitted for a device the specialization constants can all be resolved, and that the kernel never needs to be jitted for that device again?

Reply:
The [SYCL 2020 specification](https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#_specialization_constants) describes specialization constants this way:

> Device code can make use of specialization constants which represent constants whose values can be set dynamically during execution of the SYCL application. The values of these constants are fixed when a SYCL kernel function is invoked, and they do not change during the execution of the kernel. However, the application is able to set a new value for a specialization constant each time a kernel is invoked, so the values can be tuned differently for each invocation.

Reply:
OK, so the above should change to say "A constant variable where the value is not known until invocation of the SYCL kernel function."  rather than compilation.

Reply:
Updated.

Reply:
> OK, so the above should change to say "A constant variable where the value is not known until invocation of the SYCL kernel function." rather than compilation.

@kbsmith-intel, @romanovvlad, please note that this was a quote from the spec (Glossary section), so if we change it here we should also reflect that in the spec.

> Can we get clarification about exactly what constitutes compilation time of the SYCL kernel function?

I guess it depends on whether we are talking about online or offline compilation. Plus, even for offline compilation SYCL spec doesn't specify the exact formats used by the compiler:

> When a SYCL application contains SYCL kernel function objects, the SYCL implementation must provide
> an offline compilation mechanism that enables the integration of the device binaries into the SYCL application. The output of the offline compiler can be an intermediate representation, such as SPIR-V, that
> will be finalized during execution or a final device ISA.
> 
> 3.5. The SYCL platform model

> In ahead of time compilation does this mean that all specialization constants must be known at that time?

For AOT, we produce a native executable object and the idea that it don't need any further online compilation/linking before being executed. Therefore, we can only two ways of dealing with spec constants:

1. Capture some default values of them and embed them into generated binary
2. Emulate support of specialization constants without needing to recompile the binary

SYCL 2020 spec mandates us to support changing specialization constants values even for AOT-compiled programs, which means that we go with (2) here.

> In JIT compilation, does this mean the the first time the kernel is jitted for a device the specialization constants can all be resolved, and that the kernel never needs to be jitted for that device again?

This is not entirely true: it is possible that kernel needs to be jitted again for the same device, because it was launched with different values of specialization constants. See the following note from the spec:

> Implementations that support online compilation of kernel bundles will likely implement both methods of specialization constants using kernel bundles. Therefore, applications should expect that there is some overhead associated with invoking a kernel with
> new values for its specialization constants. A typical implementation records the values
> of specialization constants set via `handler::set_specialization_constant()` and remembers these values until a kernel is invoked (e.g. via `parallel_for()`). At this point, the
> implementation determines the bundle that contains the invoked kernel. If that bundle
> has already been compiled for the handler’s device and compiled with the correct values
> for the specialization constants, the kernel is scheduled for invocation. Otherwise, the
> implementation compiles the bundle before scheduling the kernel for invocation. Therefore, applications that frequently change the values of specialization constants may see
> an overhead associated with recompilation of the kernel’s bundle.
>
> 4.9.5. Specialization constants

Reply:
I will mark this as resolved, please speak up if you disagree

Reply:
I agree that this is resolved for the purposes of the design discussion.  I think the SYCL spec wording is still confusing :-).

Reply:
@kbsmith-intel, I quoted the SYCL spec's description of specialization constant above.  If you can tell me the part that seems confusing, I can update the spec.

Reply:
These statements, from 4.9.5, seem very clear.
Device code can make use of specialization constants which represent constants whose values can be set
dynamically during execution of the SYCL application. The values of these constants are fixed when a
SYCL kernel function is invoked, and they do not change during the execution of the kernel. However,
the application is able to set a new value for a specialization constant each time a kernel is invoked, so
the values can be tuned differently for each invocation.

This statement, from page 502 of the glossary, and what is quoted above in the design document is wrong/confusing:
specialization constant
A constant variable where the value is not known until compilation of the SYCL kernel function.

I think in the design document above, we should change the wording from the glossary wording to:
"SYCL 2020, 4.9.5: Device code can make use of specialization constants which represent constants whose values can be set
dynamically during execution of the SYCL application. The values of these constants are fixed when a
SYCL kernel function is invoked, and they do not change during the execution of the kernel. However,
the application is able to set a new value for a specialization constant each time a kernel is invoked, so
the values can be tuned differently for each invocation."

And I think the SYCL 2020 glossary should be updated to be consistent with 4.9.5.

Reply:
> And I think the SYCL 2020 glossary should be updated to be consistent with 4.9.5.

Oh yes, that should be updated.  I'll fix it.  Thanks.

Reply:
Thanks Greg